### PR TITLE
Update deep speech model with pure tensorflow API implementation

### PR DESCRIPTION
**Pull request by A30041839**
Hi,

This PR contains the final version of DeepSpeech 2 model implemented with tensorflow high level API. Changes included in this PR:
1) Add a README file and requirements.txt (yanhui & haoliang).
2) Code refactoring across the whole pipeline to improve code readability (yanhui).
3) Use tensorflow high level API to replace the previous Keras implementation (yanhui & haoliang)

The main reason we switched to TF API is that Kera's BatchNormalization layer could not be handled correctly by tf.keras.model_to_estimator. The issue we found is:
      a) The model_to_estimator function fails to add update_ops as a control dependency of the train_op. 
      b) Even if we write our own model_fn and add update_ops manually, with the current tf-nightly build, update_ops associated with Kera's BatchNormalization layer is not added to the tf.GraphKeys.UPDATE_OPs. So it's impossible for us to find a graceful solution to make Kera's BN layer work seamlessly with estimator.

For model evaluation, we obtain the best result from a 5-layer RNN with 800 hidden units. The network is trained on train-clean-100 and train-clean-360 dataset. At the 16000 steps, we reach WER of 0.21 (compared with mlperf pytorch 0.23).

Thanks a lot for reviewing this PR!

### CODE REVIEW

File: research/deep_speech/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+The training data is a combination of train-clean-100 and train-clean-360 (~130k
+examples in total). The validation set is dev-clean which has 2.7K lines.
+The download script will preprocess the data into three columns: wav_filename,
+wav_filesize, transcript. data/dataset.py will parse the csv file and build a
+tf.data.Dataset object to feed data. Within each training cycle (except for the
+first if sortagrad is enabled), the training data will be shuffled batch-wise.
+
+## Running Code
+
```
Comment (yhliang2018) [inline research/deep_speech/README.md]: I see you have a requirement.txt file for the third-party library dependencies. Could you add a section here to ask users to install them first? Something like:
### Install dependencies
First install shared dependencies before running the code. Issue the following command:
`pip3 install -r requirements.txt`
or
`pip install -r requirements.txt`

File: research/deep_speech/deep_speech.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 from __future__ import division
 from __future__ import print_function
 
+
```
Comment (yhliang2018) [inline research/deep_speech/deep_speech.py]: Could you add a random seed to the main loop?

File: research/deep_speech/deep_speech.py [hunk 3]
```diff
@@ -41,53 +42,86 @@
...
+def ctc_loss(args):
+  """Computes the ctc loss for the current batch of predictions."""
+  label_length, ctc_input_length, labels, logits = args
+  label_length = tf.to_int32(tf.squeeze(label_length))
+  ctc_input_length = tf.to_int32(tf.squeeze(ctc_input_length))
+  sparse_labels = tf.to_int32(
+      tf.keras.backend.ctc_label_dense_to_sparse(labels, label_length))
+  y_pred = tf.log(tf.transpose(
+      logits, perm=[1, 0, 2]) + tf.keras.backend.epsilon())
```
Comment (qlzh727) [inline research/deep_speech/deep_speech.py]: What's the reason to add a small epsilon here?

File: research/deep_speech/deep_speech.py [hunk 4]
```diff
@@ -103,45 +137,76 @@ def evaluate_model(
...
+    predictions = {
+        "classes": tf.argmax(logits, axis=2),
+        "probabilities": tf.nn.softmax(logits),
+        "logits": logits
+    }
+    return tf.estimator.EstimatorSpec(
+        mode=mode,
+        predictions=predictions)
+
+  # In training mode.
```
Comment (qlzh727) [inline research/deep_speech/deep_speech.py]: I guess you don't handle the EVAL model here, which probably should throw exception if the mode == EVALUATE.

File: research/deep_speech/deep_speech.py [hunk 5]
```diff
@@ -41,53 +42,86 @@
...
+      given dataset.
     input_fn_eval: data input function for evaluation.
 
   Returns:
     Evaluation result containing 'wer' and 'cer' as two metrics.
   """
   # Get predictions
-  predictions = estimator.predict(
-      input_fn=input_fn_eval, yield_single_examples=False)
+  predictions = estimator.predict(input_fn=input_fn_eval)
```
Comment (qlzh727) [inline research/deep_speech/deep_speech.py]: What's the reason to use predict instead of eval method here? Is that because you can generate proper eval metric here?

File: research/deep_speech/deep_speech_model.py [hunk 6]
```diff
@@ -12,183 +12,173 @@
...
-    ctc_loss = tf.keras.layers.Lambda(
-        _ctc_lambda_func, output_shape=(1,), name="ctc_loss")(
-            [y_pred, labels, ctc_input_length, label_length])
-
-    super(DeepSpeech, self).__init__(
-        inputs=[input_data, labels, input_length, label_length],
-        outputs=[ctc_input_length, ctc_loss, y_pred])
+      inputs = _rnn_layer(
+          inputs, rnn_cell, self.rnn_hidden_size, layer_counter + 1,
+          is_batch_norm, self.is_bidirectional, training=training)
```
Comment (qlzh727) [inline research/deep_speech/deep_speech_model.py]: nit: probably can skip "training=" since its not a named variable.

File: research/deep_speech/data/featurizer.py [hunk 7]
```diff
@@ -19,42 +19,66 @@
...
+  if max_freq > sample_rate / 2:
+    raise ValueError("max_freq must not be greater than half of sample rate.")
 
-  # Perform transpose to set its shape as [time_steps, feature_num_bins]
-  spectrogram = np.transpose(np.absolute(stft), (1, 0))
-  return spectrogram
+  if stride_ms > window_ms:
+    raise ValueError("Stride size must not be greater than window size.")
+
+  stride_size = int(0.001 * sample_rate * stride_ms)
```
[the remark this answers is not in the document]
Comment (A30041839) [reply research/deep_speech/data/featurizer.py]: sample_rate is particularly very large, such as 16000. Basically stride_size is number of samples per stride_ms, and should always be greater than 1. If somehow it reaches 0, I guess that's a mis-configuration and spectro-extract will probably fail.
Our default configuration for those are:
sample_rate: 16000
stride_ms: 10ms

File: research/deep_speech/decoder.py [hunk 8]
```diff
@@ -18,189 +18,78 @@
...
+  def __init__(self, labels, blank_index=28):
     """Decoder initialization.
 
     Arguments:
-      labels (string): mapping from integers to characters.
-      blank_index (int, optional): index for the blank '_' character.
-        Defaults to 0.
-      space_index (int, optional): index for the space ' ' character.
+      labels: a string specifying the speech labels for the decoder to use.
+      blank_index: an integer specifying index for the blank character.
```
[the remark this answers is not in the document]
Comment (A30041839) [reply research/deep_speech/decoder.py]: '_' appears as the last element of our vocabulary as a pure placeholder character for blank output, so this is not really meant for outputting an underscore:) To work with CTC loss, it expects that we manually insert this placeholder into our vocabulary. In the last fully connected layer, the network will learn to output a blank symbol if there is no enough meaningful features at a time frame. For example, if I utter something like 'hello' and with the utterance may contain some disconnections, this will be captured by the network as an output form like 'h-ell-o'. The special '-' sign will be trimmed by CTC loss function when computing the actual loss.

File: research/deep_speech/deep_speech.py (hunk 3 above)
Comment (A30041839) [reply research/deep_speech/deep_speech.py]: Adding an epsilon will prevent tf.log to occur NaN error if logits are too small (for example 0).

File: research/deep_speech/deep_speech.py (hunk 5 above)
Comment (A30041839) [reply research/deep_speech/deep_speech.py]: Yeah, we will use the result (logits) from predict to compute metrics ourselves. Since we haven't implemented this in eval mode, so here we will just call predict.

File: research/deep_speech/deep_speech.py (hunk 2 above)
Comment (A30041839) [reply research/deep_speech/deep_speech.py]: Discussed offline, we don't need setting random seeds as it's behavior is not fully expected.

File: research/deep_speech/README.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
+# DeepSpeech2 Model
+## Overview
+This is an implementation of the [DeepSpeech2](https://arxiv.org/pdf/1512.02595.pdf) model. Current implementation is based on the code from the authors' [DeepSpeech code](https://github.com/PaddlePaddle/DeepSpeech) and the implementation in the [MLPerf Repo](https://github.com/mlperf/reference/tree/master/speech_recognition).
+
+DeepSpeech2 is an end-to-end deep neural network for automatic speech
+recognition (ASR). It consists of 2 convolutional layers, 5 bidirectional RNN
+layers and a fully connected layer. The feature in use is linear spectrogram
+extracted from audio input. The network uses Connectionist Temporal Classification (CTC) as the loss function.
```
[the remark this answers is not in the document]
Comment (A30041839) [reply research/deep_speech/README.md]: Appending a link to paper.

File: research/deep_speech/data/dataset.py [hunk 10]
```diff
@@ -33,40 +34,36 @@ class AudioConfig(object):
...
 
-  def __init__(self, audio_config, data_path, vocab_file_path):
+  def __init__(self, audio_config, data_path, vocab_file_path, sortagrad):
     """Initialize the configs for deep speech dataset.
 
     Args:
       audio_config: AudioConfig object specifying the audio-related configs.
       data_path: a string denoting the full path of a manifest file.
       vocab_file_path: a string specifying the vocabulary file path.
+      sortagrad: a boolean to control whether sorting the audio in the first
```
[the remark this answers is not in the document]
Comment (A30041839) [reply research/deep_speech/data/dataset.py]: Add more comments:)

File: research/deep_speech/data/dataset.py [hunk 11]
```diff
@@ -178,22 +154,48 @@ def __init__(self, dataset_config):
...
+  Returns:
+    The shuffled data entries.
+  """
+  shuffled_entries = []
+  if epoch_index == 0 and sortagrad:
+    # No need to shuffle.
+    shuffled_entries = entries
+  else:
+    # Shuffle entries batch-wise.
+    max_buckets = int(math.ceil(len(entries) / batch_size))
```
[the remark this answers is not in the document]
Comment (A30041839) [reply research/deep_speech/data/dataset.py]: Sure. I choose to shuffle on full batches and just append the remaining batch into the result.

File: research/deep_speech/deep_speech.py (hunk 4 above)
Comment (A30041839) [reply research/deep_speech/deep_speech.py]: Raise error here instead.

File: research/deep_speech/data/dataset.py [hunk 12]
```diff
@@ -126,19 +122,20 @@ def _preprocess_data(file_path):
...
-    A list of tuples (audio_file_path, file_size, transcript) sorted by file
-    size.
+    A list of tuples (wav_filename, wav_filesize, transcript) sorted by
+    file_size.
   """
   tf.logging.info("Loading data set {}".format(file_path))
   with tf.gfile.Open(file_path, "r") as f:
     lines = f.read().splitlines()
-  # Skip the csv header in lines[0].
+  # Skip the tsv header in lines[0].
```
Comment (yhliang2018) [inline research/deep_speech/data/dataset.py]: what is tsv?

Comment (A30041839) [reply research/deep_speech/data/dataset.py]: tab-separated values? https://en.wikipedia.org/wiki/Tab-separated_values

File: research/deep_speech/data/dataset.py [hunk 13]
```diff
@@ -178,22 +154,48 @@ def __init__(self, dataset_config):
...
+    The shuffled data entries.
+  """
+  shuffled_entries = []
+  if epoch_index == 0 and sortagrad:
+    # No need to shuffle.
+    shuffled_entries = entries
+  else:
+    # Shuffle entries batch-wise.
+    max_buckets = int(math.ceil(len(entries) / batch_size))
+    total_buckets = [i for i in xrange(max_buckets)]
```
[the remark this answers is not in the document]
Comment (yhliang2018) [reply research/deep_speech/data/dataset.py]: Sadly, I think here we cannot guarantee that the dataset length can be dividable by batch_size.

File: research/deep_speech/data/featurizer.py [hunk 14]
```diff
@@ -23,10 +23,14 @@
...
-  """Compute the spectrograms for the input samples(waveforms)."""
+  """Compute the spectrograms for the input samples(waveforms).
+
+  More about spectrogram computation, please refer to:
+  https://en.wikipedia.org/wiki/Short-time_Fourier_transform.
+  """
   if max_freq is None:
-    max_freq = sample_rate / 2
-  if max_freq > sample_rate / 2:
+    max_freq = sample_rate // 2
```
Comment (yhliang2018) [inline research/deep_speech/data/featurizer.py]: Do we need max_freq to be integer here?

File: research/deep_speech/data/featurizer.py [hunk 15]
```diff
@@ -112,3 +116,4 @@ def __init__(self, vocab_file):
       self.index_to_token[index] = line
       self.speech_labels += line
       index += 1
+
```
Comment (yhliang2018) [inline research/deep_speech/data/featurizer.py]: nit: remove the extra blank line.

**Comment by qlzh727** [comment]
Thanks for the update. This overall looks good, and i think its good to checkin to meet the tight timeline here. We still have chance to address issue in future if needed.

File: research/deep_speech/data/featurizer.py (hunk 14 above)
Comment (A30041839) [reply research/deep_speech/data/featurizer.py]: I think so.. frequency should be treated as an integer.

File: research/deep_speech/deep_speech.py (hunk 4 above)
Comment (yhliang2018) [reply research/deep_speech/deep_speech.py]: The EVAL mode will be handled by the following codes if someone calls estimator.evaluate(), and it just has no eval metrics (which is not mandatory in eval mode). I don't think raise error is a good idea.

**Comment by A30041839** [comment]
Thanks Yanhui & Scott for your review! I appreciate that we can check-in code to meet timeline first.

File: research/deep_speech/data/dataset.py (hunk 12 above)
Comment (yhliang2018) [reply research/deep_speech/data/dataset.py]: I think csv is more general and understandable. We can use csv here to keep consistent with line 122.

File: research/deep_speech/data/featurizer.py (hunk 14 above)
Comment (yhliang2018) [reply research/deep_speech/data/featurizer.py]: But if you check line 61 of variable freqs, it's a floating number. Could you confirm this part? I think we need to make it concise.
